### PR TITLE
Adding additional tests

### PR DESCRIPTION
**Pull request by fidian**
Additional tests, as requested in #100. This includes the scenarios from #100, #102, #103.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -15,7 +15,8 @@
     "Martin Kolárik (https://kolarik.sk)",
     "Paul Miller (paulmillr.com)",
     "Tom Byrer (https://github.com/tomByrer)",
-    "(https://github.com/DianeLooney)"
+    "(https://github.com/DianeLooney)",
+    "Tyler Akins <fidian@rumkin.com> (http://rumkin.com)"
```
Comment (jonschlinkert) [inline package.json]: this will be auto-generated, thank you for adding it though

**Comment by fidian** [comment]
Oh, the instructions when I was submitting a pull request said for me to
manually add it. Perhaps those instructions should get an update?

--
Tyler Akins

**Comment by jonschlinkert** [comment]
> Perhaps those instructions should get an update?

Indeed it should. We've upgraded our tooling since then, thanks for the heads up!

File: test/extglobs.js [hunk 2]
```diff
@@ -49,6 +49,13 @@ describe('extglobs', function() {
     mm(['a.js', 'a.txt', 'a.md'], 'a.!(js)*', ['a.md', 'a.txt']);
   });
 
+  it('handles extglobs and patterns, issues 100, 102, 103', function() {
+    mm(['foo/MyDir/MyFile.jsx'], 'foo/MyDir/**/*.+(js|jsx)', ['foo/MyDir/MyFile.jsx']);
+    mm(['file.txt'], '**/!(folder).txt', ['file.txt']);
+    mm(['a/dir/file.txt'], '*/dir/**/!(folder).txt', ['a/dir/file.txt']);
+    mm(['a/b/c.txt', 'a/b/cc.txt'], '*/b/!(c).txt', ['a/b/cc.txt']);
```
Comment (jonschlinkert) [inline test/extglobs.js]: this test is incorrect. fwiw, minimatch will match this, but it shouldn't match according to bash conventions.

Comment (fidian) [reply test/extglobs.js]: Really? My version of Bash does work.

image

Comment (fidian) [reply test/extglobs.js]: I should also add, that according to the manual, "cc" is not "c".

              !(pattern-list)
                     Matches anything except one of the given patterns

Comment (jonschlinkert) [reply test/extglobs.js]: maybe I was relying on an old bash. what version are you using? this is relatively easy to change, I just want to make sure it's correct.

Comment (fidian) [reply test/extglobs.js]: This functionality is widely supported by Bash. I maintain the [Multishell Docker Container](https://hub.docker.com/r/fidian/multishell/), which has all of the available versions of Bash within. I wrote a test script and executed it in the same directory structure as what I had before. Here is the test scrpt:

```
#!/usr/bin/env bash
extra=

if ! shopt -s extglob &> /dev/null; then
    extra="$extra -extglob"
fi

if ! shopt -s globstar &> /dev/null; then
    extra="$extra -globstar"
fi

matches=(*/b/!(c).txt)

if [[ "${#matches[@]}" -eq 1 && "${matches[0]}" == "a/b/cc.txt" ]]; then
    result="pass"
else
    result="FAIL"
fi

echo "$result $BASH_VERSION: ""${matches[@]}""$extra"
```

Test results against official versions of Bash:

```
FAIL 2.02.0(4)-release: */b/!(c).txt -globstar
FAIL 2.02.1(1)-release: */b/!(c).txt -globstar
pass 2.03.0(1)-release: a/b/cc.txt -globstar
pass 2.04.0(1)-release: a/b/cc.txt -globstar
pass 2.05.0(1)-release: a/b/cc.txt -globstar
pass 2.05a.0(1)-release: a/b/cc.txt -globstar
pass 2.05b.0(1)-release: a/b/cc.txt -globstar
pass 3.00.0(2)-release: a/b/cc.txt -globstar
pass 3.00.1(2)-release: a/b/cc.txt -globstar
pass 3.00.10(2)-release: a/b/cc.txt -globstar
pass 3.00.11(2)-release: a/b/cc.txt -globstar
pass 3.00.12(2)-release: a/b/cc.txt -globstar
pass 3.00.13(2)-release: a/b/cc.txt -globstar
pass 3.00.14(2)-release: a/b/cc.txt -globstar
pass 3.00.15(2)-release: a/b/cc.txt -globstar
pass 3.00.16(2)-release: a/b/cc.txt -globstar
pass 3.00.17(2)-release: a/b/cc.txt -globstar
pass 3.00.18(2)-release: a/b/cc.txt -globstar
pass 3.00.19(2)-release: a/b/cc.txt -globstar
pass 3.00.2(2)-release: a/b/cc.txt -globstar
pass 3.00.20(2)-release: a/b/cc.txt -globstar
pass 3.00.21(2)-release: a/b/cc.txt -globstar
pass 3.00.22(3)-release: a/b/cc.txt -globstar
pass 3.00.3(2)-release: a/b/cc.txt -globstar
pass 3.00.4(2)-release: a/b/cc.txt -globstar
pass 3.00.5(2)-release: a/b/cc.txt -globstar
pass 3.00.6(2)-release: a/b/cc.txt -globstar
pass 3.00.7(2)-release: a/b/cc.txt -globstar
pass 3.00.8(2)-release: a/b/cc.txt -globstar
pass 3.00.9(2)-release: a/b/cc.txt -globstar
pass 3.1.0(2)-release: a/b/cc.txt -globstar
pass 3.1.1(2)-release: a/b/cc.txt -globstar
pass 3.1.10(2)-release: a/b/cc.txt -globstar
pass 3.1.11(2)-release: a/b/cc.txt -globstar
pass 3.1.12(2)-release: a/b/cc.txt -globstar
pass 3.1.13(2)-release: a/b/cc.txt -globstar
pass 3.1.14(2)-release: a/b/cc.txt -globstar
pass 3.1.15(2)-release: a/b/cc.txt -globstar
pass 3.1.16(2)-release: a/b/cc.txt -globstar
pass 3.1.17(2)-release: a/b/cc.txt -globstar
pass 3.1.18(2)-release: a/b/cc.txt -globstar
pass 3.1.19(2)-release: a/b/cc.txt -globstar
pass 3.1.2(2)-release: a/b/cc.txt -globstar
pass 3.1.20(2)-release: a/b/cc.txt -globstar
pass 3.1.21(2)-release: a/b/cc.txt -globstar
pass 3.1.22(2)-release: a/b/cc.txt -globstar
pass 3.1.23(2)-release: a/b/cc.txt -globstar
pass 3.1.3(2)-release: a/b/cc.txt -globstar
pass 3.1.4(2)-release: a/b/cc.txt -globstar
pass 3.1.5(2)-release: a/b/cc.txt -globstar
pass 3.1.6(2)-release: a/b/cc.txt -globstar
pass 3.1.7(2)-release: a/b/cc.txt -globstar
pass 3.1.8(2)-release: a/b/cc.txt -globstar
pass 3.1.9(2)-release: a/b/cc.txt -globstar
pass 3.2.0(2)-release: a/b/cc.txt -globstar
pass 3.2.1(2)-release: a/b/cc.txt -globstar
pass 3.2.10(2)-release: a/b/cc.txt -globstar
pass 3.2.11(2)-release: a/b/cc.txt -globstar
pass 3.2.12(2)-release: a/b/cc.txt -globstar
pass 3.2.13(2)-release: a/b/cc.txt -globstar
pass 3.2.14(2)-release: a/b/cc.txt -globstar
pass 3.2.15(2)-release: a/b/cc.txt -globstar
pass 3.2.16(2)-release: a/b/cc.txt -globstar
pass 3.2.17(2)-release: a/b/cc.txt -globstar
pass 3.2.18(2)-release: a/b/cc.txt -globstar
pass 3.2.19(2)-release: a/b/cc.txt -globstar
pass 3.2.2(2)-release: a/b/cc.txt -globstar
pass 3.2.20(2)-release: a/b/cc.txt -globstar
pass 3.2.21(2)-release: a/b/cc.txt -globstar
pass 3.2.22(2)-release: a/b/cc.txt -globstar
pass 3.2.23(2)-release: a/b/cc.txt -globstar
pass 3.2.24(2)-release: a/b/cc.txt -globstar
pass 3.2.25(2)-release: a/b/cc.txt -globstar
pass 3.2.26(2)-release: a/b/cc.txt -globstar
pass 3.2.27(2)-release: a/b/cc.txt -globstar
pass 3.2.28(2)-release: a/b/cc.txt -globstar
pass 3.2.29(2)-release: a/b/cc.txt -globstar
pass 3.2.3(2)-release: a/b/cc.txt -globstar
pass 3.2.30(2)-release: a/b/cc.txt -globstar
pass 3.2.31(2)-release: a/b/cc.txt -globstar
pass 3.2.32(2)-release: a/b/cc.txt -globstar
pass 3.2.33(2)-release: a/b/cc.txt -globstar
pass 3.2.34(2)-release: a/b/cc.txt -globstar
pass 3.2.35(2)-release: a/b/cc.txt -globstar
pass 3.2.36(2)-release: a/b/cc.txt -globstar
pass 3.2.37(2)-release: a/b/cc.txt -globstar
pass 3.2.38(3)-release: a/b/cc.txt -globstar
pass 3.2.39(3)-release: a/b/cc.txt -globstar
pass 3.2.4(2)-release: a/b/cc.txt -globstar
pass 3.2.40(3)-release: a/b/cc.txt -globstar
pass 3.2.41(3)-release: a/b/cc.txt -globstar
pass 3.2.42(3)-release: a/b/cc.txt -globstar
pass 3.2.43(3)-release: a/b/cc.txt -globstar
pass 3.2.44(3)-release: a/b/cc.txt -globstar
pass 3.2.45(3)-release: a/b/cc.txt -globstar
pass 3.2.46(3)-release: a/b/cc.txt -globstar
pass 3.2.47(3)-release: a/b/cc.txt -globstar
pass 3.2.48(3)-release: a/b/cc.txt -globstar
pass 3.2.49(3)-release: a/b/cc.txt -globstar
pass 3.2.5(2)-release: a/b/cc.txt -globstar
pass 3.2.50(3)-release: a/b/cc.txt -globstar
pass 3.2.51(3)-release: a/b/cc.txt -globstar
pass 3.2.52(3)-release: a/b/cc.txt -globstar
pass 3.2.53(2)-release: a/b/cc.txt -globstar
pass 3.2.54(3)-release: a/b/cc.txt -globstar
pass 3.2.55(2)-release: a/b/cc.txt -globstar
pass 3.2.56(3)-release: a/b/cc.txt -globstar
pass 3.2.57(2)-release: a/b/cc.txt -globstar
pass 3.2.6(2)-release: a/b/cc.txt -globstar
pass 3.2.7(2)-release: a/b/cc.txt -globstar
pass 3.2.8(2)-release: a/b/cc.txt -globstar
pass 3.2.9(2)-release: a/b/cc.txt -globstar
pass 4.0.0(2)-release: a/b/cc.txt
pass 4.0.0(2)-rc1: a/b/cc.txt
pass 4.0.1(2)-release: a/b/cc.txt
pass 4.0.10(2)-release: a/b/cc.txt
pass 4.0.11(2)-release: a/b/cc.txt
pass 4.0.12(2)-release: a/b/cc.txt
pass 4.0.13(2)-release: a/b/cc.txt
pass 4.0.14(2)-release: a/b/cc.txt
pass 4.0.15(2)-release: a/b/cc.txt
pass 4.0.16(2)-release: a/b/cc.txt
pass 4.0.17(2)-release: a/b/cc.txt
pass 4.0.18(2)-release: a/b/cc.txt
pass 4.0.19(2)-release: a/b/cc.txt
pass 4.0.2(2)-release: a/b/cc.txt
pass 4.0.20(2)-release: a/b/cc.txt
pass 4.0.21(2)-release: a/b/cc.txt
pass 4.0.22(2)-release: a/b/cc.txt
pass 4.0.23(2)-release: a/b/cc.txt
pass 4.0.24(2)-release: a/b/cc.txt
pass 4.0.25(2)-release: a/b/cc.txt
pass 4.0.26(2)-release: a/b/cc.txt
pass 4.0.27(2)-release: a/b/cc.txt
pass 4.0.28(2)-release: a/b/cc.txt
pass 4.0.29(2)-release: a/b/cc.txt
pass 4.0.3(2)-release: a/b/cc.txt
pass 4.0.30(2)-release: a/b/cc.txt
pass 4.0.31(2)-release: a/b/cc.txt
pass 4.0.32(2)-release: a/b/cc.txt
pass 4.0.33(2)-release: a/b/cc.txt
pass 4.0.34(2)-release: a/b/cc.txt
pass 4.0.35(2)-release: a/b/cc.txt
pass 4.0.36(2)-release: a/b/cc.txt
pass 4.0.37(2)-release: a/b/cc.txt
pass 4.0.38(2)-release: a/b/cc.txt
pass 4.0.39(2)-release: a/b/cc.txt
pass 4.0.4(2)-release: a/b/cc.txt
pass 4.0.40(2)-release: a/b/cc.txt
pass 4.0.41(2)-release: a/b/cc.txt
pass 4.0.42(2)-release: a/b/cc.txt
pass 4.0.43(2)-release: a/b/cc.txt
pass 4.0.44(2)-release: a/b/cc.txt
pass 4.0.5(2)-release: a/b/cc.txt
pass 4.0.6(2)-release: a/b/cc.txt
pass 4.0.7(2)-release: a/b/cc.txt
pass 4.0.8(2)-release: a/b/cc.txt
pass 4.0.9(2)-release: a/b/cc.txt
pass 4.1.0(2)-release: a/b/cc.txt
pass 4.1.1(2)-release: a/b/cc.txt
pass 4.1.10(2)-release: a/b/cc.txt
pass 4.1.11(2)-release: a/b/cc.txt
pass 4.1.12(2)-release: a/b/cc.txt
pass 4.1.13(2)-release: a/b/cc.txt
pass 4.1.14(2)-release: a/b/cc.txt
pass 4.1.15(2)-release: a/b/cc.txt
pass 4.1.16(2)-release: a/b/cc.txt
pass 4.1.17(2)-release: a/b/cc.txt
pass 4.1.2(2)-release: a/b/cc.txt
pass 4.1.3(2)-release: a/b/cc.txt
pass 4.1.4(2)-release: a/b/cc.txt
pass 4.1.5(2)-release: a/b/cc.txt
pass 4.1.6(2)-release: a/b/cc.txt
pass 4.1.7(2)-release: a/b/cc.txt
pass 4.1.8(2)-release: a/b/cc.txt
pass 4.1.9(2)-release: a/b/cc.txt
pass 4.2.0(2)-release: a/b/cc.txt
pass 4.2.1(2)-release: a/b/cc.txt
pass 4.2.10(2)-release: a/b/cc.txt
pass 4.2.11(2)-release: a/b/cc.txt
pass 4.2.12(2)-release: a/b/cc.txt
pass 4.2.13(2)-release: a/b/cc.txt
pass 4.2.14(2)-release: a/b/cc.txt
pass 4.2.15(2)-release: a/b/cc.txt
pass 4.2.16(2)-release: a/b/cc.txt
pass 4.2.17(2)-release: a/b/cc.txt
pass 4.2.18(2)-release: a/b/cc.txt
pass 4.2.19(2)-release: a/b/cc.txt
pass 4.2.2(2)-release: a/b/cc.txt
pass 4.2.20(2)-release: a/b/cc.txt
pass 4.2.21(2)-release: a/b/cc.txt
pass 4.2.22(2)-release: a/b/cc.txt
pass 4.2.23(2)-release: a/b/cc.txt
pass 4.2.24(2)-release: a/b/cc.txt
pass 4.2.25(2)-release: a/b/cc.txt
pass 4.2.26(2)-release: a/b/cc.txt
pass 4.2.27(2)-release: a/b/cc.txt
pass 4.2.28(2)-release: a/b/cc.txt
pass 4.2.29(2)-release: a/b/cc.txt
pass 4.2.3(2)-release: a/b/cc.txt
pass 4.2.30(2)-release: a/b/cc.txt
pass 4.2.31(2)-release: a/b/cc.txt
pass 4.2.32(2)-release: a/b/cc.txt
pass 4.2.33(2)-release: a/b/cc.txt
pass 4.2.34(2)-release: a/b/cc.txt
pass 4.2.35(2)-release: a/b/cc.txt
pass 4.2.36(2)-release: a/b/cc.txt
pass 4.2.37(2)-release: a/b/cc.txt
pass 4.2.38(2)-release: a/b/cc.txt
pass 4.2.39(2)-release: a/b/cc.txt
pass 4.2.4(2)-release: a/b/cc.txt
pass 4.2.40(2)-release: a/b/cc.txt
pass 4.2.41(2)-release: a/b/cc.txt
pass 4.2.42(2)-release: a/b/cc.txt
pass 4.2.43(2)-release: a/b/cc.txt
pass 4.2.44(2)-release: a/b/cc.txt
pass 4.2.45(2)-release: a/b/cc.txt
pass 4.2.46(2)-release: a/b/cc.txt
pass 4.2.47(2)-release: a/b/cc.txt
pass 4.2.48(2)-release: a/b/cc.txt
pass 4.2.49(2)-release: a/b/cc.txt
pass 4.2.5(2)-release: a/b/cc.txt
pass 4.2.50(2)-release: a/b/cc.txt
pass 4.2.51(4)-release: a/b/cc.txt
pass 4.2.52(4)-release: a/b/cc.txt
pass 4.2.53(4)-release: a/b/cc.txt
pass 4.2.6(2)-release: a/b/cc.txt
pass 4.2.7(2)-release: a/b/cc.txt
pass 4.2.8(2)-release: a/b/cc.txt
pass 4.2.9(2)-release: a/b/cc.txt
pass 4.3.0(1)-release: a/b/cc.txt
pass 4.3.1(1)-release: a/b/cc.txt
pass 4.3.10(1)-release: a/b/cc.txt
pass 4.3.11(1)-release: a/b/cc.txt
pass 4.3.12(1)-release: a/b/cc.txt
pass 4.3.13(1)-release: a/b/cc.txt
pass 4.3.14(1)-release: a/b/cc.txt
pass 4.3.15(1)-release: a/b/cc.txt
pass 4.3.16(1)-release: a/b/cc.txt
pass 4.3.17(1)-release: a/b/cc.txt
pass 4.3.18(1)-release: a/b/cc.txt
pass 4.3.19(1)-release: a/b/cc.txt
pass 4.3.2(1)-release: a/b/cc.txt
pass 4.3.20(1)-release: a/b/cc.txt
pass 4.3.21(1)-release: a/b/cc.txt
pass 4.3.22(1)-release: a/b/cc.txt
pass 4.3.23(1)-release: a/b/cc.txt
pass 4.3.24(1)-release: a/b/cc.txt
pass 4.3.25(1)-release: a/b/cc.txt
pass 4.3.26(1)-release: a/b/cc.txt
pass 4.3.27(1)-release: a/b/cc.txt
pass 4.3.28(1)-release: a/b/cc.txt
pass 4.3.29(1)-release: a/b/cc.txt
pass 4.3.3(1)-release: a/b/cc.txt
pass 4.3.30(1)-release: a/b/cc.txt
pass 4.3.31(1)-release: a/b/cc.txt
pass 4.3.32(1)-release: a/b/cc.txt
pass 4.3.33(1)-release: a/b/cc.txt
pass 4.3.34(1)-release: a/b/cc.txt
pass 4.3.35(1)-release: a/b/cc.txt
pass 4.3.36(1)-release: a/b/cc.txt
pass 4.3.37(1)-release: a/b/cc.txt
pass 4.3.38(1)-release: a/b/cc.txt
pass 4.3.39(1)-release: a/b/cc.txt
pass 4.3.4(1)-release: a/b/cc.txt
pass 4.3.40(1)-release: a/b/cc.txt
pass 4.3.41(1)-release: a/b/cc.txt
pass 4.3.42(1)-release: a/b/cc.txt
pass 4.3.43(1)-release: a/b/cc.txt
pass 4.3.44(1)-release: a/b/cc.txt
pass 4.3.45(1)-release: a/b/cc.txt
pass 4.3.46(1)-release: a/b/cc.txt
pass 4.3.47(1)-release: a/b/cc.txt
pass 4.3.48(1)-release: a/b/cc.txt
pass 4.3.5(1)-release: a/b/cc.txt
pass 4.3.6(1)-release: a/b/cc.txt
pass 4.3.7(1)-release: a/b/cc.txt
pass 4.3.8(1)-release: a/b/cc.txt
pass 4.3.9(1)-release: a/b/cc.txt
pass 4.4.0(1)-release: a/b/cc.txt
pass 4.4.0(1)-beta: a/b/cc.txt
pass 4.4.0(1)-beta2: a/b/cc.txt
pass 4.4.0(1)-rc1: a/b/cc.txt
pass 4.4.0(1)-rc2: a/b/cc.txt
pass 4.4.1(1)-release: a/b/cc.txt
pass 4.4.10(1)-release: a/b/cc.txt
pass 4.4.11(1)-release: a/b/cc.txt
pass 4.4.12(1)-release: a/b/cc.txt
pass 4.4.2(1)-release: a/b/cc.txt
pass 4.4.3(1)-release: a/b/cc.txt
pass 4.4.4(1)-release: a/b/cc.txt
pass 4.4.5(1)-release: a/b/cc.txt
pass 4.4.6(1)-release: a/b/cc.txt
pass 4.4.7(1)-release: a/b/cc.txt
pass 4.4.8(1)-release: a/b/cc.txt
pass 4.4.9(1)-release: a/b/cc.txt
```

And these are the versions of Bash used by Apple:

```
pass 3.2.57(2)-release: a/b/cc.txt -globstar
pass 3.2.57(2)-release: a/b/cc.txt -globstar
pass 2.05a.0(1)-release: a/b/cc.txt -globstar
pass 2.05b.0(1)-release: a/b/cc.txt -globstar
pass 2.05b.0(1)-release: a/b/cc.txt -globstar
pass 2.05b.0(1)-release: a/b/cc.txt -globstar
pass 2.05b.0(1)-release: a/b/cc.txt -globstar
pass 2.05b.0(1)-release: a/b/cc.txt -globstar
pass 2.05b.0(1)-release: a/b/cc.txt -globstar
pass 2.05b.0(1)-release: a/b/cc.txt -globstar
pass 3.2.17(2)-release: a/b/cc.txt -globstar
pass 3.2.17(2)-release: a/b/cc.txt -globstar
pass 3.2.48(3)-release: a/b/cc.txt -globstar
pass 3.2.48(2)-release: a/b/cc.txt -globstar
pass 3.2.53(4)-release: a/b/cc.txt -globstar
pass 3.2.48(2)-release: a/b/cc.txt -globstar
pass 3.2.53(2)-release: a/b/cc.txt -globstar
pass 3.2.51(2)-release: a/b/cc.txt -globstar
pass 3.2.53(2)-release: a/b/cc.txt -globstar
pass 3.2.53(2)-release: a/b/cc.txt -globstar
pass 3.2.57(2)-release: a/b/cc.txt -globstar
pass 3.2.57(2)-release: a/b/cc.txt -globstar
```

**Pull request merged**